### DISCUSSION TITLE
Missing layout options hGrab and vGrab

### DISCUSSION THREAD

**Discussion by srcnuzn** [eclipse-glsp/glsp#1069] [Q&A]
According to the documentation about [client-layouting ](https://eclipse.dev/glsp/documentation/clientlayouting/) there are the options `hGrab` and `vGrab`, but they seem to be missing in [GLayoutOptions](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.graph/src/org/eclipse/glsp/graph/builder/impl/GLayoutOptions.java). Also, setting the options manually does not have any effect on my `GCompartment`.

For example, the labels in the following  `GCompartment` within my node are not grabbing the vertical space but instead are compactly occupying the space in the center. Am I misunderstanding the option?

```java
new GCompartmentBuilder()
    .layout(GConstants.Layout.VBOX)
    .layoutOptions(Map.of(
            GLayoutOptions.KEY_H_ALIGN, GConstants.HAlign.LEFT,
            GLayoutOptions.KEY_V_ALIGN, GConstants.VAlign.CENTER,
            "vGrab", true))
    .add(new GLabelBuilder().text("p12").build())
    .add(new GLabelBuilder().text("p1234").build())
    .add(new GLabelBuilder().text("p123456").build())
    .build()
```

<img src="https://github.com/eclipse-glsp/glsp/assets/61967763/c8703c9e-f1ae-48f4-831e-414ff51366ae" width="40%" height="40%">

**Comment by CamilleLetavernier** [comment]
Hi,

The vGrab/hGrab options affect the element on which they are defined. In this case, the compartment that contains all 3 labels, rather than individual labels.

Unfortunately, Labels do not support layout options, so you can't apply vGrab on them directly (This is what you'd need to get the `expected result`). As a workaround, you may try wrapping each label in their own compartment, and applying vGrab on these intermediate compartments. It makes the model a lot more verbose, but it should give the expected result.

**Comment by srcnuzn** [comment]
Hey,

I tried your suggestion and with the following code, the labels have a higher vertical distance, but unfortunately they are still not dynamically grabbing the whole vertically available space. Note, that it makes no difference whether or not I add the `vGrab `option.

```java
new GCompartmentBuilder()
    .layout(GConstants.Layout.VBOX)
    .layoutOptions(Map.of(
            GLayoutOptions.KEY_H_ALIGN, GConstants.HAlign.LEFT,
            GLayoutOptions.KEY_V_ALIGN, GConstants.VAlign.CENTER,
            "vGrab", true)) // vGrab has no effect
    .add(getLabel("p12"))
    .add(getLabel("p1234"))
    .add(getLabel("p123456"))
    .build()
```

```java    
private GCompartment getLabel(final String text) {
    return new GCompartmentBuilder()
            .layout(GConstants.Layout.VBOX)
            .layoutOptions(Map.of("vGrab", true))  // vGrab has no effect
            .add(new GLabelBuilder().text(text).build())
            .build();
}
```
<img src="https://github.com/eclipse-glsp/glsp/assets/61967763/e6a41137-e75d-45bc-b269-e03dc314d9e4" width="30%" height="30%">

**Reply by rsoika** [reply]
Hi @srcnuzn , I have a very similar problem which seems to occur since I upgraded from 1.1.0-RC04 => 1.1.0-RC10. So in the moment I just wonder if you get your problem solved - and at least you are not alone ;-)
This is how it worked in the past in my project:
![Peek 2023-07-25 23-17](https://github.com/eclipse-glsp/glsp/assets/1339124/5728e816-9ad2-4dd5-be10-ad7881cdd476)

**Comment by rsoika** [comment]
Hi @CamilleLetavernier , I also believe we have a problem here. The example code from @srcnuzn,  where he is wrapping all labels in a `GCompartment` and applying the Layout options to each, does not show the expected effect. 
I run into this problem too since I updated to 1.1.0-RC10. For example now it is no longer possible to set the `HAlign.RIGHT`. 
Can you please show an example code to demonstrate how to construct such an element?

**Comment by rsoika** [comment]
@srcnuzn, finally the only solution in my case to get it working correctly, was to use a `GConstants.Layout.FREEFORM` and placing the elements manually.
For this solution you also need to react active on the event `ChangeBoundsOperation`  to recompute the position of your elements based on the new dimensions. 

This is of course more work but you finally have much more control over the layout

![Peek 2023-07-26 23-36](https://github.com/eclipse-glsp/glsp/assets/1339124/951b1d74-2c28-4696-8412-0c154d5010ff)

**Reply by srcnuzn** [reply]
Thanks for the hint! I also thought about manually layouting the ports. I will give it a try and let you know whether or not I get it done as desired.

**Comment by tortmayr** [comment]
Thanks for raising this.
We also have received bug reports from other adopters that seem to be related to this problem. We have to investigate the 
root cause and hopefully can provide a fix soon.
In the meantime I suggest to try to use the fix that @rsoika mentioned.

**Comment by rsoika** [comment]
Hi @tortmayr,
No with my new solution I observer a really strange behaviour caused by the `FREEFORM` layout. After I changed my Element impl each Time I move an element or change a route all Elements with the `FREEFORM` layout grow in its size a little bit. Look at this annimation:

![Peek 2023-07-28 17-49](https://github.com/eclipse-glsp/glsp/assets/1339124/fdb99c96-4fe9-491c-86e2-274c57cd0acb)

If I switch back to `VBOX` layout the effect is gone:

![Peek 2023-07-28 17-50](https://github.com/eclipse-glsp/glsp/assets/1339124/c1e30ad6-74be-4396-9c06-dced91ea1b80)

Do you know about this effect?

**Reply by rsoika** [reply]
@tortmayr , I found out that the reason for the strange effect was simply that one of my child components (GNode) had a position of 1,1 and graps the width and height form the parent. And this leads to the effect that the parent increases its size in each render phase by 1 px. So you can ignore my concerns regarding the FREEFORM layout. So the whole thing was just a later consequence of switching from VBOX layout to FREEFORM. I need to refactor more parts of my impl....
